### DISCUSSION TITLE
Conditional PartOf specifications

### DISCUSSION THREAD

**Discussion by andyward** [buildingSMART/IDS#341] [Ideas]
I'd like to write a specification that implements the requirement that **All spaces belonging to a BuildingStorey '01' should have their space names matching '01-.*'**  (or the converse requirement: **Spaces named like '01-.*' must be part of a BuildingStorey with Name '01'** )

I'm struggling to see how I could do this in IDS1.0. I was hoping to be able to do this with the _partOf_ facet and using **RelAggregates**  but there's no obvious way to apply a requirement to the 'child' side of the relationship. 

e.g.
```xml
<specification name="Spaces on level '01' must be named based on the level" ifcVersion="IFC2X3">
  <applicability maxOccurs="unbounded">
    <entity>
      <name>
        <simpleValue>IFCBUILDINGSTOREY</simpleValue>
      </name>
    </entity>
    <ids:attribute>
      <ids:name>
        <simpleValue>Name</simpleValue>
      </ids:name>
      <ids:value>
        <simpleValue>01</simpleValue>
      </ids:value>
    </ids:attribute>
  </applicability>
  <requirements>
    <partOf relation="IFCRELAGGREGATES">
      <entity>
        <name>
          <simpleValue>IFCSPACE</simpleValue>
        </name>
      </entity>
      <!-- and these spaces should have Name attribute matching pattern '01-.*' -->
    </partOf>
  </requirements>
</specification>
```

Any ideas? or is this a usecase to consider for a future IDS? I'd imagine similar use cases for things like ensuring components are in the correct IfcSystem (by classification, name etc).

**Comment by atomczak** [comment]
I tried this before too, and to the best of my knowledge, it is not possible with IDS1.0. While I would love 1.1 to include that, I don't think there is an easy way to solve it, other than embedding sub-facets inside partOf. If you got an idea how you would like to capture it, let's discuss with the group.

**Reply by andyward** [reply]
>  If you got an idea how you would like to capture it, let's discuss with the group.

I suppose I thought that given `partof` [already references](https://github.com/buildingSMART/IDS/blob/b12e39224c35dac92f9adff068eb0cba74134621/Schema/ids.xsd#L51) the IFC entity facet (essentially as part of the constraint), is there any reason we can't also re-use/reference other facet types (that would also constrain the child items.)

```xml
...
// applicable to Storey named '01' 
 <requirements>
    <partOf relation="IFCRELAGGREGATES">
      <entity>
        <name>
          <simpleValue>IFCSPACE</simpleValue>
        </name>
      </entity>
      <attribute instructions="Spaces names on Level 01 should have a Name starting 01-">
        <name>Name</name>
        <value>
          <xs:restriction base="xs:string">01-.*<xs:restriction>
        </value>
      </attribute>
    </partOf>
  </requirements>

```

i.e. We'd just be composing a rule from other facets. Partof is already doing this, but just for the entity facet.

**Comment by NickNisbet** [comment]
Ask Leon about Austrian proposal.

**Comment by fissim-at** [comment]
Hi,
We have published a paper suggesting embedding sub-facets into the PartOf facet for exactly this purpose. We have included the property and attribute facets in the same way as the entity facet was already included.
Please have a look: [https://doi.org/10.1016/j.dibe.2024.100560](https://doi.org/10.1016/j.dibe.2024.100560)
